## 실습 코드
* 버섯 분류 문제 풀이
* 사이킷 런 패키지의 서포트 벡터 머신 이용
* 잘 이해되지 않는 부분은 언제든 이메일 또는 카톡(개인톡, 수업 오픈 카톡방 모두 상관 없음) 주세요.

In [1]:
# pandas 모듈 추가
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC

In [2]:
# 데이터셋 로드
mushroom = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/mushroom/agaricus-lepiota.data', header=None)

In [3]:
# 데이터 앞 다섯 줄만 보기
mushroom.head()

0  1  2  3  4  5  6  7  8  9   ... 13 14 15 16 17 18 19 20 21 22
0  p  x  s  n  t  p  f  c  n  k  ...  s  w  w  p  w  o  p  k  s  u
1  e  x  s  y  t  a  f  c  b  k  ...  s  w  w  p  w  o  p  n  n  g
2  e  b  s  w  t  l  f  c  b  n  ...  s  w  w  p  w  o  p  n  n  m
3  p  x  y  w  t  p  f  c  n  n  ...  s  w  w  p  w  o  p  k  s  u
4  e  x  s  g  f  n  f  w  b  k  ...  s  w  w  p  w  o  e  n  a  g

[5 rows x 23 columns]

In [4]:
mushroom.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       8124 non-null   object
 1   1       8124 non-null   object
 2   2       8124 non-null   object
 3   3       8124 non-null   object
 4   4       8124 non-null   object
 5   5       8124 non-null   object
 6   6       8124 non-null   object
 7   7       8124 non-null   object
 8   8       8124 non-null   object
 9   9       8124 non-null   object
 10  10      8124 non-null   object
 11  11      8124 non-null   object
 12  12      8124 non-null   object
 13  13      8124 non-null   object
 14  14      8124 non-null   object
 15  15      8124 non-null   object
 16  16      8124 non-null   object
 17  17      8124 non-null   object
 18  18      8124 non-null   object
 19  19      8124 non-null   object
 20  20      8124 non-null   object
 21  21      8124 non-null   object
 22  22      8124 non-null   

데이터가 8124줄이고, 한 줄은 23 개의 속성으로 이루어짐

* 목적
  * 데이터가 문자로 이루어짐
  * 해당 데이터를 머신러닝 모델의 입력으로 넣을 부분 X와
  * 모델이 맞출 y를 구분하고
  * 입력 X에 들어가는 문자 데이터는 숫자로 변환

In [5]:
# 문자를 숫자로 변환
X = []
y = []

"""
ierrows 함수를 쓰면, mushroom이라는 데이터프레임으로부터
줄 번호 row_index와 그 줄의 데이터를 가져옴
"""
for row_index, row in mushroom.iterrows():
    # 데이터에서 첫째 칼럼은 독성의 유무를 의미함
    # 머신러닝 모델에서 예측해야하는 것
    y.append(row[0])
    row_X = []
    # 첫째 칼럼을 제외한 나머지 칼럼은 버섯의 특징을 나타냄
    for v in row[1:]:
        # 파이썬 내장함수 ord를 쓰면 인자로 넣은 문자에 해당하는 정수 유니코드를 알려줌
        row_X.append(ord(v))
    X.append(row_X)

In [6]:
# 학습 데이터와 테스트 데이터 나누기
X_train, X_test, y_train, y_test = train_test_split(X, y)

In [7]:
# 학습 및 테스트
svc = SVC()
svc.fit(X_train, y_train)

SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None, coef0=0.0,
    decision_function_shape='ovr', degree=3, gamma='scale', kernel='rbf',
    max_iter=-1, probability=False, random_state=None, shrinking=True,
    tol=0.001, verbose=False)

정확도 평균을 나타내는 함수를 써서 모델 평가
* 사용법
  * 모델.score(X, y)

In [8]:
print(f"훈련 점수: {svc.score(X_train, y_train):.3f}")
print(f"테스트 점수: {svc.score(X_test, y_test):.3f}")

훈련 점수: 0.997
테스트 점수: 0.994


1에 가까울 수록 높은 정확도